README.md
Streaming API implementation of [MessagePack](https://msgpack.org/) binary serialization format - msgpack.

[![pub package](https://img.shields.io/pub/v/s5_msgpack.svg)](https://pub.dartlang.org/packages/s5_msgpack)
[![MIT License](https://img.shields.io/badge/License-BSD3-blue.svg)](https://opensource.org/licenses/BSD-3-Clause)

## The basics

`Packer` and `Unpacker` classes provide Streaming API for serializing and deserializing data.  
`Unpacker` also provide automatic unpacking `Map` and `List` by implicitly unpacking internal items.

## Streaming API Example - Simple

For the simplest data packaging to bytes packet like this: 
```
-------------------------------------------   
| version | userId | broadcast? | message |
-------------------------------------------
```

### Packer part

```dart
import 'package:s5_msgpack/s5_msgpack.dart';
```

```dart
final p = Packer();
p.packInt(1); // packet format version
p.packInt(222); // user id
p.packBool(false); // broadcast message to others
p.packString('hi'); // user message text
final bytes = p.takeBytes(); //Uint8List
yourFunctionSendToServer(bytes); //sends [1, 204, 222, 195, 162, 104, 105]
```
> `1` encodes to 1 byte with value 1.  
> `222` encodes to 2 bytes with values [204, 222] because 222 > 127.  
> `true` encodes to 1 byte with value 195.  
> `'hi'` encodes to 3 bytes with first byte containing str length info and other 2 bytes hold symbols values.  

> For more information, refer to the [msgpack](https://msgpack.org/) documentation.  

Streaming packing process not like the way we usually pack data in json, by specifying keys.  
The current example is more like how data is packed in TCP/IP frame structure.

### Unpacker part

```dart
final List<int> rawBytes = yourFunctionReceiveFromServer(); // receive List<int> bytes from server
final u = Unpacker.fromList(rawBytes);
final version = u.unpackInt();
final userId = u.unpackInt();
final broadcast = u.unpackBool();
final message = u.unpackString();

// check values in test
expect(version, equals(1));
expect(userId, equals(222));
expect(broadcast, equals(true));
expect(message, equals('hi'));

```

## Streaming packing and automatic/implicit unpacking example - complex

```dart
final p = Packer()
  ..packListLength(10)              //pack 10 different types items to list
  ..packInt(99)
  ..packBool(true)
  ..packString('hi')
  ..packNull()                      // explicitly pack null
  ..packString(null)                // implicitly any type can accept null 
  ..packBinary(<int>[104, 105])     // hi codes
  ..packListLength(2)               // pack 2 elements list ['elem1',3.14]
  ..packString('elem1')             // list[0]
  ..packDouble(3.14)                // list[1]
  ..packString('continue to pack other elements')
  ..packMapLength(1)                // map {'key1':false}
  ..packString('key1')              // map key
  ..packBool(false)                 // map value 
  ..packInt(9223372036854775807);   // next root list item (map ended)

final bytes = p.takeBytes();
final u = Unpacker(bytes);
// Unpack by the same sequential/streaming way 
// or implicitly/automatically
final l = u.unpackList(); //List<Object>
print(l);
// [99, true, hi, null, null, [104, 105], [elem1, 3.14], continue to pack other elements, {key1: false}, 9223372036854775807]
```

`List<Object>` items explicitly casted to corresponding types when using.

## NOTES

### Packing Maps and Lists
* firstly, pack Map or List header length 
* secondly, manually pack all items - that's all 

Only need put length header before packing items 

After packing all items no need to stop or finish or end this map / list

```dart
final list = ['i1','i2'];
final map = {'k1': 11, 'k2': 22};
final p = Packer();
p.packListLength(list.length);
list.forEach(p.packString);
p.packMapLength(map.length);
map.forEach((key, v) {
  p.packString(key);
  p.packInt(v);
});
final bytes = p.takeBytes();
```

### More examples

More examples can be found in:
* `test/messagepack_test.dart`
* `example/example.dart`

### Don't use Packer after calling .takeBytes()

Internally it cleans up underlying `_builder`, so further behaviour will be useless.  
Call `.takeBytes()` only once. It returns `Uint8List` of packed bytes.  
After that don't continue to use `Packer` instance (don't call `.packXXX()` methods and `.takeBytes()` anymore).  
Instead create new `Packer` instance.

### Motivation for creating this package

No other packages available that give streaming API for processing data 

## Contributing

If you have advice how to improve library code or making it lighter or blazingly faster - don't hesitate to open an issue or pull request!


CHANGELOG.md
## [0.3.0]
* Keep Uint8List

## [0.2.0]

* Migrate to null safety
* Require Dart SDK 2.12 or later

## [0.1.2]

* Web support

## [0.1.1]

* New functions for _implicitly/automatically_ unpack `Map` and `List` items without using **Streaming API** directly.
* * `unpackList()`
* * `unpackMap()`
* Refactored tests and added more tests
* Improved documentation

## [0.1.0] 

* Few examples 
* More tests
* Improved documentation 

## [0.1.0-dev.1]  

* Pre release tested and functional streaming API


test/messagepack_test.dart
import 'dart:convert';
import 'dart:math';

import 'package:s5_msgpack/s5_msgpack.dart';
import 'package:test/test.dart';

int? packUnpackInt(int? v, {bool negative = false}) {
  final p = Packer();
  if (negative)
    p.packInt(v);
  else
    p.packInt(v);
  final u = Unpacker(p.takeBytes());
  return u.unpackInt();
}

double? packUnpackDouble(double? v) {
  final p = Packer();
  p.packDouble(v);
  final u = Unpacker(p.takeBytes());
  return u.unpackDouble();
}

String? packUnpackString(String? v) {
  final p = Packer();
  p.packString(v);
  final u = Unpacker(p.takeBytes());
  return u.unpackString();
}

final ints = [
  9223372036854775807,
  1581367842777,
  0,
  1581367842,
  127,
  5,
  128,
  null,
  128,
];

final negativeInts = ints.map<int?>((e) => e == null ? null : -1 * e).toList()
  ..add(-9223372036854775808);

final doubles = [
  0.0,
  null,
  1.1,
  null,
  null,
  12345678.12345668,
  null,
  -1.1,
  2.56,
  -34.567890,
  -23456.54,
  null,
  -1.0
];

final strs = [
  'hi',
  null,
  null,
  'Повседневная практика показывает, что постоянное информационно-пропагандистское обеспечение нашей деятельности в значительной степени обуславливает создание модели развития. Не следует, однако забывать, что реализация намеченных плановых заданий требуют от нас анализа форм развития. Не следует, однако забывать, что новая модель организационной деятельности требуют определения и уточнения модели развития. Равным образом постоянный количественный рост и сфера нашей активности требуют от нас анализа системы обучения кадров, соответствует насущным потребностям.',
  'mediul length string',
  '12345678901234567890123456789012345678901234567890',
  '123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890',
  null,
  'str containing  кириллицу',
  'hi'
];

void main() {
  test('int sequentially increasing [independent]', () {
    for (int v = -65536; v < 65536; v++) {
      final r = packUnpackInt(v, negative: v < 0);
      expect(r, equals(v));
    }
  });

  test('Positive int [independent]', () {
    for (final v in ints) expect(packUnpackInt(v), equals(v));
  });

  test('Positive int [dependent]', () {
    final p = Packer();
    ints.forEach(p.packInt);
    final u = Unpacker(p.takeBytes());
    for (final v in ints) expect(u.unpackInt(), equals(v));
  });

  test('Negative int [independent]', () {
    for (final v in negativeInts)
      expect(packUnpackInt(v, negative: true), equals(v));
  });

  test('Negative int [dependent]', () {
    final p = Packer();
    negativeInts.forEach(p.packInt);
    final u = Unpacker(p.takeBytes());
    for (final v in negativeInts) expect(u.unpackInt(), equals(v));
  });

  test('Double [independent]', () {
    for (final v in doubles) expect(packUnpackDouble(v), equals(v));
  });

  test('Double [dependent]', () {
    final p = Packer();
    doubles.forEach(p.packDouble);
    final u = Unpacker(p.takeBytes());
    for (final v in doubles) expect(u.unpackDouble(), equals(v));
  });

  test('String [independent]', () {
    for (final v in strs) expect(packUnpackString(v), equals(v));
  });

  test('String [dependent]', () {
    final p = Packer();
    strs.forEach(p.packString);
    final u = Unpacker(p.takeBytes());
    for (final v in strs) expect(u.unpackString(), equals(v));
  });

  List<MapEntry<int, dynamic>> randomiseData() {
    final rands = <MapEntry<int, dynamic>>[];
    MapEntry<int, dynamic> me(int k, dynamic v) => MapEntry<int, dynamic>(k, v);
    final r = Random();

    dynamic nullOr(dynamic transform(int v)) {
      return r.nextInt(10) == 0 ? null : transform(r.nextInt(4294967296));
    }

    for (int i = 0; i < 200; i++) {
      final t = r.nextInt(5);
      if (t == 0) rands.add(me(t, null));
      if (t == 1) rands.add(me(t, nullOr((v) => v % 3 == 0)));
      if (t == 2) rands.add(me(t, nullOr((v) => v)));
      if (t == 3) rands.add(me(t, nullOr((v) => v * -1)));
      if (t == 4)
        rands.add(me(t,
            nullOr((v) => r.nextDouble() * v * (r.nextInt(2) == 1 ? -1 : 1))));
      if (t == 5) rands.add(me(t, nullOr((v) => strs[r.nextInt(strs.length)])));
    }
    return rands;
  }

  void pack(Packer p, MapEntry<int, dynamic> e) {
    if (e.key == 0) p.packNull();
    if (e.key == 1) p.packBool(e.value as bool?);
    if (e.key == 2) p.packInt(e.value as int?);
    if (e.key == 3) p.packInt(e.value as int?);
    if (e.key == 4) p.packDouble(e.value as double?);
    if (e.key == 5) p.packString(e.value as String?);
  }

  dynamic unpack(Unpacker u, int type) {
    if (type == 0) return unpack(u, Random().nextInt(4) + 1);
    if (type == 1) return u.unpackBool();
    if (type == 2) return u.unpackInt();
    if (type == 3) return u.unpackInt();
    if (type == 4) return u.unpackDouble();
    if (type == 5) return u.unpackString();
  }

  test('Random different [independent]', () {
    final l = randomiseData();
    for (final e in l) {
      final p = Packer();
      pack(p, e);
      final u = Unpacker(p.takeBytes());
      expect(unpack(u, e.key), equals(e.value));
    }
  });

  test('Random different [dependent]', () {
    final l = randomiseData();
    final p = Packer();
    for (final e in l) {
      pack(p, e);
    }
    final u = Unpacker(p.takeBytes());
    for (int i = 0; i < l.length; i++) {
      expect(unpack(u, l[i].key), equals(l[i].value));
    }
  });

  test('Iterable [dependent]', () {
    final list = [
      'value1 short',
      'value2 medium medium',
      'value3 with кириллица кириллица кириллица',
      'value4 with long long text that so long long long long long long long long long long',
      'value5',
      'v6'
    ];
    final p = Packer();
    p.packNull();
    p.packListLength(null);
    p.packListLength(0);
    p.packListLength(list.length);
    list.forEach(p.packString);
    p.packBool(true);
    p.packListLength(list.length);
    list.forEach(p.packString);
    p.packInt(3);
    final u = Unpacker(p.takeBytes());
    expect(u.unpackListLength(), equals(0));
    expect(u.unpackListLength(), equals(0));
    expect(u.unpackListLength(), equals(0));
    final length = u.unpackListLength();
    for (int i = 0; i < length; i++) {
      expect(u.unpackString(), equals(list[i]));
    }
    expect(u.unpackBool(), equals(true));
    final length2 = u.unpackListLength();
    for (int i = 0; i < length2; i++) {
      expect(u.unpackString(), equals(list[i]));
    }
    expect(u.unpackInt(), equals(3));
  });

  test('Iterable different types [dependent]', () {
    final list = [
      -65999,
      -34855,
      -128,
      -31,
      -1,
      0,
      null,
      124,
      1581367842,
      1581367842777,
      null,
      'short string',
      'string with medium length',
      true,
      null,
      false,
      -3.14,
      0.5,
      36.6,
      null,
      100.0,
    ];
    final p = Packer();
    p.packListLength(list.length);
    for (int i = 0; i < 10; i++) {
      p.packInt(list[i] as int?);
    }
    for (int i = 10; i < 13; i++) {
      p.packString(list[i] as String?);
    }
    for (int i = 13; i < 16; i++) {
      p.packBool(list[i] as bool?);
    }
    for (int i = 16; i < 21; i++) {
      p.packDouble(list[i] as double?);
    }
    final u = Unpacker(p.takeBytes());
    expect(u.unpackListLength(), equals(list.length));
    for (int i = 0; i < 10; i++) {
      expect(u.unpackInt(), equals(list[i]));
    }
    for (int i = 10; i < 13; i++) {
      expect(u.unpackString(), equals(list[i]));
    }
    for (int i = 13; i < 16; i++) {
      expect(u.unpackBool(), equals(list[i]));
    }
    for (int i = 16; i < 21; i++) {
      expect(u.unpackDouble(), equals(list[i]));
    }
  });

  test('Map [dependent]', () {
    final map = {
      1: 11,
      2: 22,
      3: 33,
    };
    final p = Packer();
    p.packNull();
    p.packMapLength(null);
    p.packMapLength(0);
    p.packMapLength(map.length);
    for (int i = 0; i < map.length; i++) {
      p.packInt(i);
      p.packInt(map[i]);
    }
    p.packBool(true);
    p.packMapLength(map.length);
    for (int i = 0; i < map.length; i++) {
      p.packInt(i);
      p.packInt(map[i]);
    }
    p.packInt(3);
    final u = Unpacker(p.takeBytes());
    expect(u.unpackMapLength(), equals(0));
    expect(u.unpackMapLength(), equals(0));
    expect(u.unpackMapLength(), equals(0));
    final length = u.unpackMapLength();
    for (int i = 0; i < length; i++) {
      expect(u.unpackInt(), equals(i));
      expect(u.unpackInt(), equals(map[i]));
    }
    expect(u.unpackBool(), equals(true));
    final length2 = u.unpackMapLength();
    for (int i = 0; i < length2; i++) {
      expect(u.unpackInt(), equals(i));
      expect(u.unpackInt(), equals(map[i]));
    }
    expect(u.unpackInt(), equals(3));
  });

  test('Binary [dependent]', () {
    final bytes1 = utf8.encode('hi');
    final bytes2 = utf8.encode(
        'Значимость этих проблем настолько очевидна, что укрепление и развитие структуры представляет собой интересный эксперимент проверки систем массового участия. Задача организации, в особенности же постоянное информационно-пропагандистское обеспечение нашей деятельности обеспечивает широкому кругу (специалистов) участие в формировании систем массового участия. Товарищи! постоянный количественный рост и сфера нашей активности требуют определения и уточнения соответствующий условий активизации. Задача организации, в особенности же консультация с широким активом влечет за собой процесс внедрения и модернизации форм развития. Равным образом начало повседневной работы по формированию позиции позволяет оценить значение соответствующий условий активизации.');
    final empty = <int>[];
    final p = Packer();
    p.packBinary(null);
    p.packBinary(empty);
    p.packBinary(bytes1);
    p.packBool(true);
    p.packBinary(bytes2);
    p.packInt(3);
    final u = Unpacker(p.takeBytes());
    expect(u.unpackBinary(), equals(empty));
    expect(u.unpackBinary(), equals(empty));
    expect(u.unpackBinary(), equals(bytes1));
    expect(u.unpackBool(), equals(true));
    expect(u.unpackBinary(), equals(bytes2));
    expect(u.unpackInt(), equals(3));
  });

  test('Manual int example [dependent]', () {
    final p = Packer();
    p.packInt(1); // packet format version
    p.packInt(222); // user id
    p.packBool(true); // broadcast message to others
    p.packString('hi'); // user message text
    final bytes = p.takeBytes(); //Uint8List
    //yourFunctionSendToServer(bytes); //sends [1, 204, 222, 195, 162, 104, 105]
    // 1 encodes to 1 byte with value 1
    // 222 encodes to 2 bytes with values [204, 222] because 222 > 127
    // true encodes to 1 byte with value 195
    // 'hi' encodes to 3 bytes with first byte containing str length info and other 2 bytes hold symbols values
    // For more information, refer to the msgpack documentation
    print(bytes);
    //receive bytes from server
    final u = Unpacker(bytes);
    final version = u.unpackInt();
    final userId = u.unpackInt();
    final broadcast = u.unpackBool();
    final message = u.unpackString();
    expect(version, equals(1));
    expect(userId, equals(222));
    expect(broadcast, equals(true));
    expect(message, equals('hi'));
  });

  test('Map Iterable example [dependent]', () {
    final list = ['i1', 'i2'];
    final map = {'k1': 11, 'k2': 22};
    final p = Packer();
    p.packListLength(list.length);
    list.forEach(p.packString);
    p.packMapLength(map.length);
    map.forEach((key, v) {
      p.packString(key);
      p.packInt(v);
    });
    final bytes = p.takeBytes();
    Unpacker(bytes);
  });

  test('Map automatically', () {
    final list = ['i1', 'i2'];
    final p = Packer();
    p.packListLength(list.length);
    list.forEach(p.packString);
    final bytes = p.takeBytes();
    final u = Unpacker(bytes);
    final l = u.unpackList();
    print(l);
  });

  test('Different types example [dependent]', () {
    final p = Packer()
      ..packListLength(10) //pack 10 different types items to list
      ..packInt(99)
      ..packBool(true)
      ..packString('hi')
      ..packNull()
      ..packString(null)
      ..packBinary(<int>[104, 105]) // hi codes
      ..packListLength(2) // pack 2 elements list ['elem1',3.14]
      ..packString('elem1')
      ..packDouble(3.14)
      ..packString('continue to pack other elements')
      ..packMapLength(1) //map {'key1':false}
      ..packString('key1')
      ..packBool(false)
      ..packInt(9223372036854775807);

    final bytes = p.takeBytes();
    final u = Unpacker(bytes);
    final l = u.unpackList();
    print(l);
  });
}


lib/src/unpacker.dart
import 'dart:convert';
import 'dart:typed_data';

/// Streaming API for unpacking (deserializing) data from msgpack binary format.
///
/// unpackXXX methods returns value if it exist, or `null`.
/// Throws [FormatException] if value is not an requested type,
/// but in that case throwing exception not corrupt internal state,
/// so other unpackXXX methods can be called after that.
class Unpacker {
  /// Manipulates with provided [Uint8List] to sequentially unpack values.
  /// Use [Unpaker.fromList()] to unpack raw `List<int>` bytes.
  Unpacker(this._list) : _d = ByteData.view(_list.buffer, _list.offsetInBytes);

  ///Convenient
  Unpacker.fromList(List<int> l) : this(Uint8List.fromList(l));

  final Uint8List _list;
  final ByteData _d;
  int _offset = 0;

  final _strCodec = const Utf8Codec();

  /// Unpack value if it exist. Otherwise returns `null`.
  ///
  /// Throws [FormatException] if value is not a bool,
  bool? unpackBool() {
    final b = _d.getUint8(_offset);
    bool? v;
    if (b == 0xc2) {
      v = false;
      _offset += 1;
    } else if (b == 0xc3) {
      v = true;
      _offset += 1;
    } else if (b == 0xc0) {
      v = null;
      _offset += 1;
    } else {
      throw _formatException('bool', b);
    }
    return v;
  }

  /// Unpack value if it exist. Otherwise returns `null`.
  ///
  /// Throws [FormatException] if value is not an integer,
  int? unpackInt() {
    final b = _d.getUint8(_offset);
    int? v;
    if (b <= 0x7f || b >= 0xe0) {
      /// Int value in fixnum range [-32..127] encoded in header 1 byte
      v = _d.getInt8(_offset);
      _offset += 1;
    } else if (b == 0xcc) {
      v = _d.getUint8(++_offset);
      _offset += 1;
    } else if (b == 0xcd) {
      v = _d.getUint16(++_offset);
      _offset += 2;
    } else if (b == 0xce) {
      v = _d.getUint32(++_offset);
      _offset += 4;
    } else if (b == 0xcf) {
      v = _d.getUint64(++_offset);
      _offset += 8;
    } else if (b == 0xd0) {
      v = _d.getInt8(++_offset);
      _offset += 1;
    } else if (b == 0xd1) {
      v = _d.getInt16(++_offset);
      _offset += 2;
    } else if (b == 0xd2) {
      v = _d.getInt32(++_offset);
      _offset += 4;
    } else if (b == 0xd3) {
      v = _d.getInt64(++_offset);
      _offset += 8;
    } else if (b == 0xc0) {
      v = null;
      _offset += 1;
    } else {
      throw _formatException('integer', b);
    }
    return v;
  }

  /// Unpack value if it exist. Otherwise returns `null`.
  ///
  /// Throws [FormatException] if value is not a Double.
  double? unpackDouble() {
    final b = _d.getUint8(_offset);
    double? v;
    if (b == 0xca) {
      v = _d.getFloat32(++_offset);
      _offset += 4;
    } else if (b == 0xcb) {
      v = _d.getFloat64(++_offset);
      _offset += 8;
    } else if (b == 0xc0) {
      v = null;
      _offset += 1;
    } else {
      throw _formatException('double', b);
    }
    return v;
  }

  /// Unpack value if it exist. Otherwise returns `null`.
  ///
  /// Empty
  /// Throws [FormatException] if value is not a String.
  String? unpackString() {
    final b = _d.getUint8(_offset);
    if (b == 0xc0) {
      _offset += 1;
      return null;
    }
    int len;

    /// fixstr 101XXXXX stores a byte array whose len is upto 31 bytes:
    if (b & 0xE0 == 0xA0) {
      len = b & 0x1F;
      _offset += 1;
    } else if (b == 0xc0) {
      _offset += 1;
      return null;
    } else if (b == 0xd9) {
      len = _d.getUint8(++_offset);
      _offset += 1;
    } else if (b == 0xda) {
      len = _d.getUint16(++_offset);
      _offset += 2;
    } else if (b == 0xdb) {
      len = _d.getUint32(++_offset);
      _offset += 4;
    } else {
      throw _formatException('String', b);
    }
    final data =
        Uint8List.view(_list.buffer, _list.offsetInBytes + _offset, len);
    _offset += len;
    return _strCodec.decode(data);
  }

  /// Unpack [List.length] if packed value is an [List] or `null`.
  ///
  /// Encoded in msgpack packet null or 0 length unpacks to 0 for convenience.
  /// Items of the [List] must be unpacked manually with respect to returned `length`
  /// Throws [FormatException] if value is not an [List].
  int unpackListLength() {
    final b = _d.getUint8(_offset);
    int len;
    if (b & 0xF0 == 0x90) {
      /// fixarray 1001XXXX stores an array whose length is upto 15 elements:
      len = b & 0xF;
      _offset += 1;
    } else if (b == 0xc0) {
      len = 0;
      _offset += 1;
    } else if (b == 0xdc) {
      len = _d.getUint16(++_offset);
      _offset += 2;
    } else if (b == 0xdd) {
      len = _d.getUint32(++_offset);
      _offset += 4;
    } else {
      throw _formatException('List length', b);
    }
    return len;
  }

  /// Unpack [Map.length] if packed value is an [Map] or `null`.
  ///
  /// Encoded in msgpack packet null or 0 length unpacks to 0 for convenience.
  /// Items of the [Map] must be unpacked manually with respect to returned `length`
  /// Throws [FormatException] if value is not an [Map].
  int unpackMapLength() {
    final b = _d.getUint8(_offset);
    int len;
    if (b & 0xF0 == 0x80) {
      /// fixmap 1000XXXX stores a map whose length is upto 15 elements
      len = b & 0xF;
      _offset += 1;
    } else if (b == 0xc0) {
      len = 0;
      _offset += 1;
    } else if (b == 0xde) {
      len = _d.getUint16(++_offset);
      _offset += 2;
    } else if (b == 0xdf) {
      len = _d.getUint32(++_offset);
      _offset += 4;
    } else {
      throw _formatException('Map length', b);
    }
    return len;
  }

  /// Unpack value if packed value is binary or `null`.
  ///
  /// Encoded in msgpack packet null unpacks to [List] with 0 length for convenience.
  /// Throws [FormatException] if value is not a binary.
  Uint8List unpackBinary() {
    final b = _d.getUint8(_offset);
    int len;
    if (b == 0xc4) {
      len = _d.getUint8(++_offset);
      _offset += 1;
    } else if (b == 0xc0) {
      len = 0;
      _offset += 1;
    } else if (b == 0xc5) {
      len = _d.getUint16(++_offset);
      _offset += 2;
    } else if (b == 0xc6) {
      len = _d.getUint32(++_offset);
      _offset += 4;
    } else {
      throw _formatException('Binary', b);
    }
    final data =
        Uint8List.view(_list.buffer, _list.offsetInBytes + _offset, len);
    _offset += len;
    return data;
  }

  Object? _unpack() {
    final b = _d.getUint8(_offset);
    if (b <= 0x7f ||
        b >= 0xe0 ||
        b == 0xcc ||
        b == 0xcd ||
        b == 0xce ||
        b == 0xcf ||
        b == 0xd0 ||
        b == 0xd1 ||
        b == 0xd2 ||
        b == 0xd3) {
      return unpackInt();
    } else if (b == 0xc0 || b == 0xc2 || b == 0xc3) {
      return unpackBool(); //null included
    } else if (b == 0xca || b == 0xcb) {
      return unpackDouble();
    } else if ((b & 0xE0) == 0xA0 ||
        b == 0xc0 ||
        b == 0xd9 ||
        b == 0xda ||
        b == 0xdb) {
      return unpackString();
    } else if (b == 0xc4 || b == 0xc5 || b == 0xc6) {
      return unpackBinary();
    } else if ((b & 0xF0) == 0x90 || b == 0xdc || b == 0xdd) {
      return unpackList();
    } else if ((b & 0xF0) == 0x80 || b == 0xde || b == 0xdf) {
      return unpackMap();
    } else {
      throw _formatException('Unknown', b);
    }
  }

  /// Automatically unpacks `bytes` to [List] where items has corresponding data types.
  ///
  /// Return types declared as [Object] instead of `dynamic` for safety reasons.
  /// You need explicitly cast to proper types. And in case with [Object]
  /// compiler checks will force you to do it whereas with `dynamic` it will not.
  List<Object?> unpackList() {
    final length = unpackListLength();
    return List.generate(length, (_) => _unpack());
  }

  /// Automatically unpacks `bytes` to [Map] where key and values has corresponding data types.
  ///
  /// Return types declared as [Object] instead of `dynamic` for safety reasons.
  /// You need explicitly cast to proper types. And in case with [Object]
  /// compiler checks will force you to do it whereas with `dynamic` it will not.
  Map<Object?, Object?> unpackMap() {
    final length = unpackMapLength();
    return {for (var i = 0; i < length; i++) _unpack(): _unpack()};
  }

  Exception _formatException(String type, int b) => FormatException(
      'Try to unpack $type value, but it\'s not an $type, byte = $b');
}


example/example.dart
import 'package:s5_msgpack/s5_msgpack.dart';

void simple() {
  final p = Packer();
  p.packInt(1);
  p.packInt(2);
  final bytes = p.takeBytes(); //Uint8List
  print(bytes);

  final u = Unpacker(bytes);
  final n1 = u.unpackInt();
  final n2 = u.unpackInt();
  print('unpacked n1=$n1 n2=$n2');
}

void differentTypesSimple() {
  final p = Packer();
  p.packInt(1);
  p.packBool(true);
  final bytes = p.takeBytes(); //Uint8List
  print(bytes);

  final u = Unpacker(bytes);
  print(u.unpackInt());
  print(u.unpackBool());
}

void iterableAndMap() {
  final list = ['i1', 'i2'];
  final map = {'k1': 11, 'k2': 22};
  final p = Packer();
  p.packListLength(list.length);
  list.forEach(p.packString);
  p.packMapLength(map.length);
  map.forEach((key, v) {
    p.packString(key);
    p.packInt(v);
  });
  final bytes = p.takeBytes();

  final u = Unpacker(bytes);
  final listLength = u.unpackListLength();
  for (int i = 0; i < listLength; i++) {
    print(u.unpackString());
  }
  final mapLength = u.unpackMapLength();
  for (int i = 0; i < listLength; i++) {
    print(u.unpackString());
    print(u.unpackInt());
  }
}

void differentTypesComplex() {
  final p = Packer()
    ..packInt(99)
    ..packBool(true)
    ..packString('hi')
    ..packNull()
    ..packString(null)
    ..packBinary(<int>[104, 105]) // hi codes
    ..packListLength(2) // pack 2 elements list ['elem1',3.14]
    ..packString('elem1')
    ..packDouble(3.14)
    ..packString('continue to pack other elements')
    ..packMapLength(2) //map {'key1':false, 'key2',3.14}
    ..packString('key1') //pack key1
    ..packBool(false) //pack value1
    ..packString('key12') //pack key1
    ..packDouble(3.13); //pack value1

  final bytes = p.takeBytes();
  final u = Unpacker(bytes);
  //Unpack the same sequential/streaming way
}
